File: packages/ruby/test/behavior_test.rb
# frozen_string_literal: true

# Normative behavioral suite — spec/behavior-vectors.json (SPECIFICATION.md
# [N-47], [N-49]).
#
# The scenarios are data. The runner below is the only language-specific part,
# which is what stops the ten ports from drifting apart the way ten hand-written
# suites did. Nothing here re-derives a case: if a behavior is not in the vector
# file, it belongs in engine_test.rb instead.
#
# Run: ruby -Ilib test/behavior_test.rb
require 'minitest/autorun'
require 'json'
require 'nebula_token'

# Executes the published scenarios against an injected clock and a store that
# can be told to lose one compare-and-set.
class BehaviorRunner
  # The vectors live at <repo root>/spec. Found by walking up from this file, so
  # the package neither hardcodes an absolute path nor vendors a stale copy.
  def self.spec_dir
    dir = File.expand_path(__dir__)
    loop do
      candidate = File.join(dir, 'spec')
      return candidate if File.file?(File.join(candidate, 'behavior-vectors.json'))

      parent = File.dirname(dir)
      raise "spec/behavior-vectors.json not found in any parent of #{__dir__}" if parent == dir

      dir = parent
    end
  end

  def self.load_vectors
    JSON.parse(File.read(File.join(spec_dir, 'behavior-vectors.json'), encoding: 'UTF-8')).freeze
  end

  # Conditions this runtime satisfies. A Ruby String is bytes plus an encoding
  # tag, so it can hold the UTF-8 spelling of an unpaired surrogate (invalid
  # Unicode) just as a JavaScript or Java string can — the scenario applies.
  SATISFIED_CONDITIONS = ['runtime-admits-invalid-unicode-strings'].freeze

  # 32 zero bytes, canonically encoded: well formed, and never the real secret.
  FORGED_VERIFIER = ('A' * NebulaToken::VERIFIER_CHARS).freeze
  FORGED_SELECTOR = ('A' * NebulaToken::SELECTOR_CHARS).freeze
  # U+D800 has no legal UTF-8 encoding; these are its CESU-8 bytes, tagged UTF-8.
  LONE_SURROGATE = "\xED\xA0\x80".dup.force_encoding(Encoding::UTF_8).freeze

  # A divergence from a published scenario is raised as Minitest::Assertion
  # ITSELF, never as a subclass of it. Minitest's StatisticsReporter buckets
  # results with `results.group_by { |r| r.failure.class }` and then counts
  # `aggregate[Assertion].size` — an exact class match. A subclass therefore
  # lands in a bucket that is counted as neither a failure nor an error, and the
  # summary line reads "0 failures, 0 errors" for a run that diverged from the
  # normative vectors. See the regression guard in BehaviorTest.
  SCENARIO_FAILURE = Minitest::Assertion

  TokenBinding = Struct.new(:token, :family_id, :expires_at)
  RunOutcome = Struct.new(:executed, :skipped)
  Skipped = Struct.new(:id, :condition)

  # Wraps the reference store so a scenario can force one compare-and-set to lose.
  class ControllableStore
    include NebulaToken::RefreshTokenStore

    attr_reader :inner

    def initialize
      @inner = NebulaToken::MemoryRefreshTokenStore.new
      @fail_next = {}
    end

    def fail_next_cas(method)
      @fail_next[method] = true
    end

    def find_by_selector(selector) = @inner.find_by_selector(selector)
    def insert(record) = @inner.insert(record)
    def revoke_family(family_id) = @inner.revoke_family(family_id)
    def revoke_user(user_id) = @inner.revoke_user(user_id)

    def mark_rotated(selector, from_status, rotated_at, replaced_by_selector)
      return false if consume('markRotated')

      @inner.mark_rotated(selector, from_status, rotated_at, replaced_by_selector)
    end

    def revoke_if_active(selector)
      return false if consume('revokeIfActive')

      @inner.revoke_if_active(selector)
    end

    private

    def consume(method)
      @fail_next.delete(method) ? true : false
    end
  end

  def initialize(vectors)
    @vectors = vectors
  end

  def self.applicable?(scenario)
    scenario['condition'].nil? || SATISFIED_CONDITIONS.include?(scenario['condition'])
  end

  # Execute every applicable scenario. Raises on the first divergence.
  def run_all
    outcome = RunOutcome.new([], [])
    @vectors['scenarios'].each do |scenario|
      unless self.class.applicable?(scenario)
        outcome.skipped << Skipped.new(scenario['id'], scenario['condition'])
        next
      end
      run_scenario(scenario)
      outcome.executed << scenario['id']
    end
    outcome
  end

  def run_scenario(scenario)
    @scenario = scenario
    cfg = @vectors['defaults'].merge(scenario['config'] || {})
    @store = ControllableStore.new
    @bindings = {}
    @issued_secrets = []
    @device_ids = []
    @now = cfg['now']
    @cfg = cfg
    @engine = build_engine(cfg['peppers'], cfg['activeKid'])

    scenario['steps'].each_with_index { |step, index| run_step(step, index) }
  end

  private

  def build_engine(kids, active_kid)
    NebulaToken::Engine.new(
      peppers: kids.to_h { |kid| [kid, @vectors['peppers'].fetch(kid)] },
      active_kid: active_kid,
      store: @store,
      absolute_ttl_seconds: @cfg['absoluteTtlSeconds'],
      idle_ttl_seconds: @cfg['idleTtlSeconds'],
      reuse_grace_seconds: @cfg['reuseGraceSeconds'],
      clock: -> { @now }
    )
  end

  # rubocop:disable Metrics/CyclomaticComplexity
  def run_step(step, index)
    case step['op']
    when 'issue' then op_issue(step, index)
    when 'refresh' then op_refresh(step, index)
    when 'revokeToken' then op_revoke_token(step, index)
    when 'revokeFamilyOf' then op_revoke_family_of(step, index)
    when 'revokeUser' then op_revoke_user(step, index)
    when 'advance' then @now += step['seconds']
    when 'reconfigure' then @engine = build_engine(step['peppers'], step['activeKid'])
    when 'failNextCas' then @store.fail_next_cas(step['method'])
    when 'expectStatusCounts' then op_expect_status_counts(step, index)
    when 'expectNoRawSecrets' then op_expect_no_raw_secrets(index)
    else fail_step(index, "unknown op #{step['op'].inspect}")
    end
  end
  # rubocop:enable Metrics/CyclomaticComplexity

  def op_issue(step, index)
    device_id = device_of(step)
    result = @engine.issue(step['userId'], device_id)
    expect = step['expect']
    fail_step(index, 'expected issue to fail') if expect && expect['ok'] == false
    check_success(result, expect, index)
    bind(step, result)
    @device_ids << device_id unless device_id.nil? || device_id.empty?
  end

  def op_refresh(step, index)
    expect = step['expect']
    result = @engine.refresh(resolve_token(step, index), device_of(step))
    wants_success = expect.nil? || expect['ok'] == true ||
                    (!expect.key?('ok') && !expect.key?('error'))

    if wants_success
      fail_step(index, "expected success, got #{result.error}") unless result.ok?
      check_success(result, expect, index)
      bind(step, result)
    else
      fail_step(index, "expected #{expect['error']}, got success") if result.ok?
      unless result.error == expect['error']
        fail_step(index, "expected #{expect['error']}, got #{result.error}")
      end
      check_attribution(result, expect, index)
    end
  end

  # [N-39] attribution, tri-state. `true` demands the field, `false` demands its
  # ABSENCE — the exclusion list (MALFORMED, UNKNOWN_KID, NOT_FOUND) is a
  # requirement too, and a truthy-only check could never observe it. A key that
  # is absent asserts nothing. In Ruby a result that resolved no record signals
  # it with nil, so nil is the absence this reads.
  def check_attribution(result, expect, index)
    return if expect.nil?

    if expect.key?('hasUserId') && !result.user_id.nil? != expect['hasUserId']
      fail_step(index, "expected user_id #{expect['hasUserId'] ? 'present' : 'absent'} ([N-39])")
    end
    return unless expect.key?('hasFamilyId') && !result.family_id.nil? != expect['hasFamilyId']

    fail_step(index, "expected family_id #{expect['hasFamilyId'] ? 'present' : 'absent'} ([N-39])")
  end

  def op_revoke_token(step, index)
    expect = step['expect']
    result = @engine.revoke_token(resolve_token(step, index))

    if expect && expect['ok'] == false
      fail_step(index, "expected #{expect['error']}, got success") if result.ok?
      unless result.error == expect['error']
        fail_step(index, "expected #{expect['error']}, got #{result.error}")
      end
      # [N-39] governs every failure result, revoke_token's included.
      check_attribution(result, expect, index)
    else
      fail_step(index, "expected success, got #{result.error}") unless result.ok?
      expect_revoked(expect, result.revoked, index)
    end
  end

  def op_revoke_family_of(step, index)
    bound = @bindings.fetch(step['of']) { fail_step(index, "unknown binding #{step['of'].inspect}") }
    expect_revoked(step['expect'], @engine.revoke_family(bound.family_id), index)
  end

  def op_revoke_user(step, index)
    expect_revoked(step['expect'], @engine.revoke_all_for_user(step['userId']), index)
  end

  def op_expect_status_counts(step, index)
    actual = { 'active' => 0, 'rotated' => 0, 'revoked' => 0 }
    @store.inner.all.each { |row| actual[row.status.to_s] += 1 }
    step['counts'].each do |status, want|
      next if actual[status] == want

      fail_step(index, "expected #{want} #{status}, got #{actual[status]} (#{actual})")
    end
  end

  def op_expect_no_raw_secrets(index)
    # to_h, not inspect: TokenRecord#inspect redacts the hashes ([N-14]), and a
    # redacted dump would make this assertion vacuous.
    dump = JSON.generate(@store.inner.all.map(&:to_h))
    @issued_secrets.each do |secret|
      fail_step(index, 'a raw verifier reached the store ([N-14])') if dump.include?(secret)
    end
    @device_ids.each do |device_id|
      fail_step(index, 'a raw device identifier reached the store ([N-14])') if dump.include?(device_id)
    end
  end

  def expect_revoked(expect, actual, index)
    return unless expect && expect.key?('revoked')
    return if actual == expect['revoked']

    fail_step(index, "expected #{expect['revoked']} revoked, got #{actual}")
  end

  def resolve_token(step, index)
    ref = step['token'] || {}
    return ref['literal'] if ref.key?('literal')

    fail_step(index, 'step has no token reference') unless ref.key?('ref')
    bound = @bindings.fetch(ref['ref']) { fail_step(index, "unknown binding #{ref['ref'].inspect}") }
    return bound.token if ref['forge'].nil?

    parts = bound.token.split('.')
    case ref['forge']
    when 'verifier' then parts[3] = FORGED_VERIFIER
    when 'unknownKid' then parts[1] = 'zz'
    when 'unknownSelector' then parts[2] = FORGED_SELECTOR
    else fail_step(index, "unknown forge #{ref['forge'].inspect}")
    end
    parts.join('.')
  end

  # Absence and the empty string are different values ([N-25]), so a missing
  # key must not collapse into "".
  def device_of(step)
    return LONE_SURROGATE if step['deviceIdKind'] == 'lone-surrogate'

    step['deviceId']
  end

  def check_success(result, expect, index)
    return if expect.nil?

    if expect.key?('generation') && result.generation != expect['generation']
      fail_step(index, "expected generation #{expect['generation']}, got #{result.generation}")
    end
    if expect.key?('kid') && (kid = result.token.split('.')[1]) != expect['kid']
      fail_step(index, "expected kid #{expect['kid']}, got #{kid}")
    end
    if expect.key?('sameFamilyAs') && result.family_id != @bindings.fetch(expect['sameFamilyAs']).family_id
      fail_step(index, 'family_id changed across rotation')
    end
    if expect.key?('sameExpiresAtAs')
      other = @bindings.fetch(expect['sameExpiresAtAs']).expires_at
      fail_step(index, "absolute deadline moved: #{other} -> #{result.expires_at}") if result.expires_at != other
    end
    return unless expect['idleEqualsExpires'] && result.idle_expires_at != result.expires_at

    fail_step(index, "idle_expires_at #{result.idle_expires_at} should be clamped to #{result.expires_at}")
  end

  def bind(step, result)
    @bindings[step['bind']] = TokenBinding.new(result.token, result.family_id, result.expires_at) if step['bind']
    @issued_secrets << result.token.split('.')[3]
  end

  def fail_step(index, message)
    requirements = (@scenario['requirements'] || []).join(', ')
    raise SCENARIO_FAILURE, "[#{@scenario['id']}] step #{index} (#{requirements}): #{message}"
  end
end

class BehaviorTest < Minitest::Test
  VECTORS = BehaviorRunner.load_vectors

  def test_spec_version_matches_the_published_vectors
    assert_equal VECTORS['spec_version'], NebulaToken::SPEC_VERSION
  end

  def test_every_published_scenario_is_executed_or_reported_as_skipped
    outcome = BehaviorRunner.new(VECTORS).run_all

    # [N-48]: a runner that silently iterated nothing must not report success.
    refute_empty outcome.executed, 'no scenario was executed'
    assert_equal VECTORS['counts']['scenarios'], outcome.executed.length + outcome.skipped.length,
                 'every published scenario must be either executed or explicitly skipped'
    assert_operator outcome.executed.length, :>=, VECTORS['counts']['unconditional'],
                    'every unconditional scenario must be executed'

    # Ruby Strings can hold invalid Unicode, so no scenario is inapplicable here.
    skipped = outcome.skipped.map { |s| "#{s.id} (#{s.condition})" }
    assert_empty skipped, "scenarios skipped: #{skipped.join(', ')}"
  end

  # [N-48] is about counting cases; this is about counting FAILURES. Minitest
  # tallies failures by exact class (`group_by { |r| r.failure.class }`), so a
  # runner that raised a SUBCLASS of Minitest::Assertion would print
  # "0 failures, 0 errors" for a run that diverged from the normative vectors —
  # a green-looking log for a red suite. A published conformance claim rests on
  # that line, so the runner's failure class is asserted, not assumed.
  def test_a_diverging_scenario_is_counted_as_a_failure
    scenario = {
      'id' => 'meta-divergence-probe',
      'requirements' => ['N-48'],
      'steps' => [
        { 'op' => 'issue', 'userId' => 'u1', 'bind' => 'a' },
        # A freshly issued token rotates; demanding REUSE_DETECTED must diverge.
        { 'op' => 'refresh', 'token' => { 'ref' => 'a' }, 'expect' => { 'ok' => false, 'error' => 'REUSE_DETECTED' } }
      ]
    }
    error = assert_raises(Minitest::Assertion) { BehaviorRunner.new(VECTORS).run_scenario(scenario) }
    assert_equal Minitest::Assertion, error.class,
                 'the runner must raise Minitest::Assertion itself: a subclass is tallied as ' \
                 'neither a failure nor an error, so a red run prints "0 failures, 0 errors"'
    assert_includes error.message, 'meta-divergence-probe'
  end

  # One test method per scenario, so a failure names the scenario it came from
  # and each runs against a fresh store and clock.
  VECTORS['scenarios'].each do |scenario|
    define_method("test_scenario_#{scenario['id'].tr('-', '_')}") do
      unless BehaviorRunner.applicable?(scenario)
        skip "runtime does not satisfy condition #{scenario['condition'].inspect}"
      end

      BehaviorRunner.new(VECTORS).run_scenario(scenario)
      # The runner raises on divergence; reaching here is the assertion.
      pass
    end
  end
end

File: packages/elixir/test/behavior_test.exs
defmodule NebulaToken.BehaviorVectors.ControllableStore do
  @moduledoc """
  `NebulaToken.MemoryStore` plus the one lever the vectors need: `failNextCas`,
  which makes the next call to a named compare-and-set report "did not apply"
  exactly once ([N-17], [N-18]).

  It is deliberately not a mock — every other call goes to the real store, so a
  scenario still observes real record states.
  """
  @behaviour NebulaToken.Store

  alias NebulaToken.MemoryStore

  def start_link do
    {:ok, inner} = MemoryStore.start_link()
    {:ok, control} = Agent.start_link(fn -> %{} end)
    {:ok, %{inner: inner, control: control}}
  end

  @doc "Arm a one-shot CAS failure. `method` is the spelling used by the vectors."
  def fail_next_cas(%{control: control}, method),
    do: Agent.update(control, &Map.put(&1, method, true))

  @doc "Every record currently stored. Not part of the store contract."
  def all(%{inner: inner}), do: MemoryStore.all(inner)

  @impl true
  def find_by_selector(%{inner: inner}, selector),
    do: MemoryStore.find_by_selector(inner, selector)

  @impl true
  def insert(%{inner: inner}, record), do: MemoryStore.insert(inner, record)

  @impl true
  def mark_rotated(store, selector, from_status, rotated_at, replaced_by_selector) do
    if armed?(store, "markRotated") do
      {:ok, false}
    else
      MemoryStore.mark_rotated(store.inner, selector, from_status, rotated_at, replaced_by_selector)
    end
  end

  @impl true
  def revoke_if_active(store, selector) do
    if armed?(store, "revokeIfActive") do
      {:ok, false}
    else
      MemoryStore.revoke_if_active(store.inner, selector)
    end
  end

  @impl true
  def revoke_family(%{inner: inner}, family_id), do: MemoryStore.revoke_family(inner, family_id)

  @impl true
  def revoke_user(%{inner: inner}, user_id), do: MemoryStore.revoke_user(inner, user_id)

  defp armed?(%{control: control}, method) do
    Agent.get_and_update(control, fn state ->
      {Map.get(state, method, false), Map.delete(state, method)}
    end)
  end
end

defmodule NebulaToken.BehaviorVectors.Runner do
  @moduledoc """
  Runner for the normative behavioral suite, `spec/behavior-vectors.json`
  ([N-47] §9.2, [N-49]).

  The scenarios are data. This module is the only language-specific part, which
  is what stops the ten ports from drifting apart the way ten hand-written
  suites did. Its op vocabulary is the `runner.ops` block of the vector file.
  """

  import ExUnit.Assertions

  alias NebulaToken.BehaviorVectors.ControllableStore
  alias NebulaToken.{Engine, Failure}

  @vectors NebulaToken.SpecVectors.load("behavior-vectors.json")

  # Conditions from `runner.conditions` that this runtime satisfies.
  #
  # `runtime-admits-invalid-unicode-strings` is NOT among them. `String.t()` on
  # the BEAM is by definition a UTF-8 binary, and an unpaired surrogate has no
  # UTF-8 encoding, so an Elixir string cannot hold one — the vector file itself
  # scopes the condition to "JavaScript, Java, C#, Dart and Python", excluding
  # UTF-8-only string types.
  #
  # The underlying [N-12] requirement is not skipped: engine_test.exs proves it
  # against <<0xED, 0xA0, 0x80>>, the WTF-8 byte sequence an unpaired surrogate
  # would encode to, which String.valid?/1 rejects and which the engine must
  # treat as a binding failure rather than an exception.
  @satisfied_conditions []

  # 32 zero bytes, canonically encoded: well-formed, and never the real secret.
  @forged_verifier String.duplicate("A", 43)
  @forged_selector String.duplicate("A", 22)

  def vectors, do: @vectors
  def scenarios, do: @vectors["scenarios"]
  def satisfied_conditions, do: @satisfied_conditions

  def satisfied?(nil), do: true
  def satisfied?(condition), do: condition in @satisfied_conditions

  @doc """
  Execute every scenario in the file and report what **actually ran** ([N-48]).

  The two lists are accumulated by running the scenarios, never derived from the
  file's own counts. That is the whole point: a runner that computes
  `executed = total - skipped` reports a number it did not earn, and stays green
  after someone narrows the loop it iterates. Here, iterating a subset — or
  nothing at all — makes `executed + skipped` disagree with `counts.scenarios`
  and fails the suite, which is what the requirement asks for.
  """
  @spec run_all() :: %{executed: [String.t()], skipped: [{String.t(), String.t()}]}
  def run_all do
    outcome =
      Enum.reduce(scenarios(), %{executed: [], skipped: []}, fn scenario, acc ->
        if satisfied?(scenario["condition"]) do
          run_scenario(scenario)
          %{acc | executed: [scenario["id"] | acc.executed]}
        else
          %{acc | skipped: [{scenario["id"], scenario["condition"]} | acc.skipped]}
        end
      end)

    %{executed: Enum.reverse(outcome.executed), skipped: Enum.reverse(outcome.skipped)}
  end

  @doc "Execute one scenario by id. Fails on the first divergence."
  def run(id) do
    scenario = Enum.find(scenarios(), &(&1["id"] == id)) || flunk("unknown scenario #{id}")
    run_scenario(scenario)
  end

  defp run_scenario(scenario) do
    config = Map.merge(@vectors["defaults"], scenario["config"] || %{})
    {:ok, store} = ControllableStore.start_link()
    {:ok, clock} = Agent.start_link(fn -> config["now"] end)

    state = %{
      scenario: scenario,
      config: config,
      store: store,
      clock: clock,
      engine: build(config, store, clock, config["peppers"], config["activeKid"]),
      bindings: %{},
      # Raw verifier parts of every token handed out, and every non-empty device
      # identifier presented, for `expectNoRawSecrets`.
      secrets: [],
      device_ids: MapSet.new()
    }

    scenario["steps"]
    |> Enum.with_index()
    |> Enum.reduce(state, fn {step, index}, acc -> run_step(acc, step, index) end)

    :ok
  end

  defp build(config, store, clock, kids, active_kid) do
    Engine.new(
      peppers: Map.new(kids, fn kid -> {kid, @vectors["peppers"][kid]} end),
      active_kid: active_kid,
      store_mod: ControllableStore,
      store: store,
      absolute_ttl_seconds: config["absoluteTtlSeconds"],
      idle_ttl_seconds: config["idleTtlSeconds"],
      reuse_grace_seconds: config["reuseGraceSeconds"],
      clock: fn -> Agent.get(clock, & &1) end
    )
  end

  # ── Ops ───────────────────────────────────────────────────────────────────

  defp run_step(state, %{"op" => "issue"} = step, index) do
    expect = step["expect"] || %{}
    device_id = device_of(state, step, index)

    refute expect["ok"] == false, context(state, index, "no scenario expects issue to fail")

    result = Engine.issue(state.engine, step["userId"], device_id)
    check_success(state, result, expect, index)

    state
    |> bind(step["bind"], result)
    |> track_secret(result.token)
    |> track_device(device_id)
  end

  defp run_step(state, %{"op" => "refresh"} = step, index) do
    expect = step["expect"] || %{}
    token = resolve_token(state, step["token"], index)
    result = Engine.refresh(state.engine, token, device_of(state, step, index))

    if expects_success?(expect) do
      case result do
        {:ok, success} ->
          check_success(state, success, expect, index)
          state |> bind(step["bind"], success) |> track_secret(success.token)

        other ->
          flunk(context(state, index, "expected success, got #{inspect(other)}"))
      end
    else
      check_failure(state, result, expect, index)
      state
    end
  end

  defp run_step(state, %{"op" => "revokeToken"} = step, index) do
    expect = step["expect"] || %{}
    token = resolve_token(state, step["token"], index)
    result = Engine.revoke_token(state.engine, token)

    if expect["ok"] == false do
      # The same check as the refresh branch, attribution included: [N-39]
      # governs every failure result, and revokeToken resolves its record at
      # step 3 before proving the verifier at step 4 ([N-36]).
      check_failure(state, result, expect, index)
    else
      case result do
        {:ok, success} -> check_revoked(state, success.revoked, expect, index)
        other -> flunk(context(state, index, "expected success, got #{inspect(other)}"))
      end
    end

    state
  end

  defp run_step(state, %{"op" => "revokeFamilyOf"} = step, index) do
    binding = fetch_binding(state, step["of"], index)
    revoked = Engine.revoke_family(state.engine, binding.family_id)
    check_revoked(state, revoked, step["expect"] || %{}, index)
    state
  end

  defp run_step(state, %{"op" => "revokeUser"} = step, index) do
    revoked = Engine.revoke_all_for_user(state.engine, step["userId"])
    check_revoked(state, revoked, step["expect"] || %{}, index)
    state
  end

  defp run_step(state, %{"op" => "advance"} = step, _index) do
    Agent.update(state.clock, &(&1 + step["seconds"]))
    state
  end

  defp run_step(state, %{"op" => "reconfigure"} = step, _index) do
    # A NEW engine over the SAME store: this is a pepper rotation, not a reset.
    %{
      state
      | engine: build(state.config, state.store, state.clock, step["peppers"], step["activeKid"])
    }
  end

  defp run_step(state, %{"op" => "failNextCas"} = step, _index) do
    ControllableStore.fail_next_cas(state.store, step["method"])
    state
  end

  defp run_step(state, %{"op" => "expectStatusCounts"} = step, index) do
    actual = state.store |> ControllableStore.all() |> Enum.frequencies_by(& &1.status)

    for {status, expected} <- step["counts"] do
      got = Map.get(actual, String.to_existing_atom(status), 0)

      assert got == expected,
             context(state, index, "expected #{expected} #{status}, got #{got} (#{inspect(actual)})")
    end

    state
  end

  defp run_step(state, %{"op" => "expectNoRawSecrets"}, index) do
    dump =
      state.store
      |> ControllableStore.all()
      |> inspect(limit: :infinity, printable_limit: :infinity)

    for secret <- state.secrets do
      refute String.contains?(dump, secret),
             context(state, index, "a raw verifier reached the store ([N-14])")
    end

    for device_id <- state.device_ids do
      refute String.contains?(dump, device_id),
             context(state, index, "a raw device identifier reached the store ([N-14])")
    end

    state
  end

  defp run_step(state, step, index),
    do: flunk(context(state, index, "unknown op #{inspect(step["op"])}"))

  # ── Expectations ──────────────────────────────────────────────────────────

  # Mirrors the reference runner: a step with neither `ok` nor `error` expects
  # success.
  defp expects_success?(expect),
    do: expect["ok"] == true or (expect["ok"] == nil and expect["error"] == nil)

  defp check_success(state, result, expect, index) do
    if expect["generation"] != nil do
      assert result.generation == expect["generation"],
             context(state, index, "expected generation #{expect["generation"]}, got #{result.generation}")
    end

    if expect["kid"] != nil do
      kid = result.token |> String.split(".") |> Enum.at(1)
      assert kid == expect["kid"], context(state, index, "expected kid #{expect["kid"]}, got #{kid}")
    end

    if expect["sameFamilyAs"] != nil do
      other = fetch_binding(state, expect["sameFamilyAs"], index)

      assert result.family_id == other.family_id,
             context(state, index, "family_id changed across rotation")
    end

    if expect["sameExpiresAtAs"] != nil do
      other = fetch_binding(state, expect["sameExpiresAtAs"], index)

      assert result.expires_at == other.expires_at,
             context(state, index, "absolute deadline moved: #{other.expires_at} -> #{result.expires_at}")
    end

    if expect["idleEqualsExpires"] == true do
      assert result.idle_expires_at == result.expires_at,
             context(state, index, "idle_expires_at #{result.idle_expires_at} should be clamped to #{result.expires_at}")
    end

    :ok
  end

  defp check_failure(state, result, expect, index) do
    case result do
      {:error, %Failure{} = failure} ->
        assert Atom.to_string(failure.code) == expect["error"],
               context(state, index, "expected #{expect["error"]}, got #{failure.code}")

        check_attribution(state, failure, expect, index)

      other ->
        flunk(context(state, index, "expected #{expect["error"]}, got #{inspect(other)}"))
    end
  end

  # [N-39] attribution, tri-state. `true` demands the field, `false` demands its
  # ABSENCE — the exclusion list (:MALFORMED, :UNKNOWN_KID, :NOT_FOUND) is a
  # requirement too, and a truthy-only check could never observe it. A key the
  # scenario omits asserts nothing.
  #
  # `%NebulaToken.Failure{}` always carries both keys, so "absent" here is nil.
  # check_failure/4 is shared by the refresh and revokeToken branches, which is
  # what makes the two agree: [N-39] governs every failure result, not refresh
  # alone.
  defp check_attribution(state, %Failure{} = failure, expect, index) do
    if expect["hasUserId"] != nil do
      assert (failure.user_id != nil) == expect["hasUserId"],
             context(state, index, "expected user_id #{presence(expect["hasUserId"])} ([N-39])")
    end

    if expect["hasFamilyId"] != nil do
      assert (failure.family_id != nil) == expect["hasFamilyId"],
             context(state, index, "expected family_id #{presence(expect["hasFamilyId"])} ([N-39])")
    end
  end

  defp presence(true), do: "present"
  defp presence(false), do: "absent"

  defp check_revoked(state, actual, expect, index) do
    if expect["revoked"] != nil do
      assert actual == expect["revoked"],
             context(state, index, "expected #{expect["revoked"]} revoked, got #{actual}")
    end
  end

  # ── Helpers ───────────────────────────────────────────────────────────────

  defp resolve_token(_state, %{"literal" => literal}, _index), do: literal

  defp resolve_token(state, %{"ref" => ref} = token, index) do
    bound = fetch_binding(state, ref, index)

    case token["forge"] do
      nil -> bound.token
      "verifier" -> replace_part(bound.token, 3, @forged_verifier)
      "unknownKid" -> replace_part(bound.token, 1, "zz")
      "unknownSelector" -> replace_part(bound.token, 2, @forged_selector)
      other -> flunk(context(state, index, "unknown forge #{inspect(other)}"))
    end
  end

  defp resolve_token(state, other, index),
    do: flunk(context(state, index, "step has no usable token reference: #{inspect(other)}"))

  defp replace_part(token, position, replacement) do
    token |> String.split(".") |> List.replace_at(position, replacement) |> Enum.join(".")
  end

  defp device_of(state, step, index) do
    case step["deviceIdKind"] do
      nil ->
        step["deviceId"]

      kind ->
        flunk(context(state, index, "deviceIdKind #{inspect(kind)} is not supported by this runtime"))
    end
  end

  defp fetch_binding(state, name, index) do
    Map.get(state.bindings, name) ||
      flunk(context(state, index, "unknown binding #{inspect(name)}"))
  end

  defp bind(state, nil, _result), do: state

  defp bind(state, name, result) do
    binding = %{
      token: result.token,
      family_id: result.family_id,
      expires_at: result.expires_at
    }

    %{state | bindings: Map.put(state.bindings, name, binding)}
  end

  defp track_secret(state, token) do
    %{state | secrets: [token |> String.split(".") |> Enum.at(3) | state.secrets]}
  end

  defp track_device(state, device_id) when device_id in [nil, ""], do: state
  defp track_device(state, device_id), do: %{state | device_ids: MapSet.put(state.device_ids, device_id)}

  defp context(state, index, message) do
    requirements = Enum.join(state.scenario["requirements"] || [], ", ")
    "[#{state.scenario["id"]}] step #{index} (#{requirements}): #{message}"
  end
end

defmodule NebulaToken.BehaviorVectorsTest do
  @moduledoc """
  Executes every scenario in `spec/behavior-vectors.json` ([N-47], [N-49]).

  One ExUnit test per scenario, generated from the file at compile time, so a
  divergence names the scenario that diverged. Conditional scenarios this
  runtime does not satisfy are tagged `:skip` and reported by id.
  """
  use ExUnit.Case, async: true

  alias NebulaToken.BehaviorVectors.Runner

  @vectors Runner.vectors()

  test "the published counts describe the file that was executed ([N-48])" do
    scenarios = @vectors["scenarios"]

    assert @vectors["spec_version"] == NebulaToken.spec_version()
    assert scenarios != []
    assert length(scenarios) == @vectors["counts"]["scenarios"]

    assert Enum.count(scenarios, &is_nil(&1["condition"])) == @vectors["counts"]["unconditional"]
  end

  test "every applicable scenario is really executed, and skips are reported by id ([N-48])" do
    # `outcome` is accumulated while the scenarios run, so the numbers asserted
    # below are the numbers earned. Deriving them from the file instead would
    # let a runner that iterates a subset — or zero cases — still report the
    # published count, which [N-48] calls a conformance failure, not a pass.
    outcome = Runner.run_all()
    executed = length(outcome.executed)
    total = length(@vectors["scenarios"])

    assert executed + length(outcome.skipped) == @vectors["counts"]["scenarios"],
           "every published scenario must be either executed or explicitly skipped"

    assert executed >= @vectors["counts"]["unconditional"],
           "every unconditional scenario must be executed"

    assert outcome.executed == Enum.uniq(outcome.executed), "a scenario ran twice"

    # A runner MUST execute every unconditional scenario; only a scenario
    # carrying a `condition` this runtime does not satisfy may be skipped.
    assert Enum.all?(outcome.skipped, fn {_id, condition} -> condition != nil end)

    report =
      Enum.map(outcome.skipped, fn {id, condition} ->
        "\n  skipped #{id} (condition: #{condition})"
      end)

    IO.puts([
      "\n[behavior-vectors] executed #{executed}/#{total} scenarios",
      report,
      if(outcome.skipped == [], do: "\n  (none skipped)", else: ""),
      "\n"
    ])
  end

  for scenario <- @vectors["scenarios"] do
    id = scenario["id"]
    name = "#{id}: #{scenario["title"]}"

    if Runner.satisfied?(scenario["condition"]) do
      test name do
        Runner.run(unquote(id))
      end
    else
      @tag skip: "runtime does not satisfy condition: #{scenario["condition"]}"
      test name do
        Runner.run(unquote(id))
      end
    end
  end
end
